### PR TITLE
feat(clients): use hardware id for device verification

### LINKED ISSUE #6837: Use hardware ID for device verification
Make a best effort to try to read the hardware Unique ID for all platforms to send to the portal before falling back to the stored device ID.

Furthermore, we will not hash this id before sending it to the portal so it's easier for customers to match between the ID seen in the portal and the one on the client.

Also, update the docs to match the new behavior.

### PR DESCRIPTION
**Pull request by conectado**
We want to associate additional device information for the device verification, these are all parameters that tries to uniquely identify the hardware.

For that reason we read system information and send it as part of the query params to the portal, that way the portal can store this when device is verified and match against that later on.

These are the parameters according to each platform:

|Platform|Query Field|Field Meaning|
|-----|----|-----|
|MacOS|`device_serial`|Hardware's Serial|
|MacOS| `device_uuid`|Hardware's UUID|
|iOS|`identifier_for_vendor`| Identifier for vendor, resets only on uninstall/install|
|Android|`firebase_installation_id`| Firebase installation ID, resets only on uninstall/install|
|Windows|`device_serial`|Motherboard's Serial|
|Linux|`device_serial`|Motherboard's Serial|

Fixes #6837

### CODE REVIEW

**Comment by conectado** [comment]
@jamilbk this is practically ready but we should decide if we keep the fallback or maybe skip smbios completely?

**Comment by conectado** [comment]
Before merging I also still need to test on iOS

**Comment by conectado** [comment]
Also to note here, after merging this PR even when we fallback, due to no longer hashing the id, this will un-verify all new clients, so this might be worth a minor version change?

**Comment by jamilbk** [comment]
> Also to note here, after merging this PR even when we fallback, due to no longer hashing the id, this will un-verify all new clients, so this might be worth a minor version change?

I don't think many are using this feature yet, so I'll just let the customer know directly.

File: website/src/components/Changelog/Android.tsx
Comment (thomaseizinger) [inline website/src/components/Changelog/Android.tsx]:
```suggestion
          Uses the Firebase Installation ID for device verification.
```

File: website/src/components/Changelog/Apple.tsx
Comment (thomaseizinger) [inline website/src/components/Changelog/Apple.tsx]:
```suggestion
          Uses the hardware's UUID on MacOS and vendor ID on iOS for device verification.
```

File: website/src/components/Changelog/GUI.tsx
Comment (thomaseizinger) [inline website/src/components/Changelog/GUI.tsx]:
```suggestion
          Tries to use motherboard's hardware ID for device verification.
```

File: website/src/components/Changelog/Headless.tsx
Comment (thomaseizinger) [inline website/src/components/Changelog/Headless.tsx]:
```suggestion
          Tries to use motherboard's hardware ID for device verification.
```

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt
Comment (thomaseizinger) [inline kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt]: Can we read that earlier? I.e. when Firezone first starts and let it crash if it doesn't work?

**Review by thomaseizinger — COMMENTED** [review]
Overall looks good. From the issue it sounds like there are more changes coming though :)

File: rust/phoenix-channel/src/login_url.rs
Comment (thomaseizinger) [inline rust/phoenix-channel/src/login_url.rs]: 1. Do we really want to pass around string => string maps? Why not make this typed, like a `struct`?
2. I'd like to discuss whether this information shouldn't go into the `User-Agent` instead (https://github.com/firezone/firezone/issues/6837#issuecomment-2384216972). That is where we have it currently.

Comment (conectado) [reply rust/phoenix-channel/src/login_url.rs]: 1. I was thinking that string => string maps would give us the most flexibility as this information will come from different places such as clients with different fields
2. that seems to be overloading the user-agent, the device id is the closest to this which is called `external_id` in the query params. Also it's easier to parse and add fields there rather than the user-agent which doesn't have a fixed structure.

Comment (thomaseizinger) [reply rust/phoenix-channel/src/login_url.rs]: > I was thinking that string => string maps would give us the most flexibility as this information will come from different places such as clients with different fields

I mean, yeah. With that argument we don't need to be using any types anywhere :)

At the very least, we should make an enum for all the keys but I'd prefer a struct that models out this information, perhaps all query params that we can set?

Comment (conectado) [reply rust/phoenix-channel/src/login_url.rs]: Using enums sounds good. But if we use a `DeviceInfo` struct it will have all `Option` fields which we would have to check one by one to add to the query params.

Comment (thomaseizinger) [reply rust/phoenix-channel/src/login_url.rs]: > But if we use a `DeviceInfo` struct it will have all `Option` fields which we would have to check one by one to add to the query params.

What be so bad? At the moment we have a lot of `Option`s as parameters :man_shrugging:

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt
Comment (conectado) [reply kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt]: This can fail in certain very edge cases and I would rather avoid crashing then

**Comment by conectado** [comment]
@thomaseizinger  Ready for another round of reviews

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt
Comment (thomaseizinger) [inline kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelService.kt:362]: This has gotten a bit verbose which was the main reason I suggested to do it elsewhere. I don't know how long it takes to fetch the Firebase ID, I guess once it is cached locally it shouldn't delay the connect much?

File: rust/connlib/clients/apple/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/apple/src/lib.rs:197]:
```suggestion
        let device_info = serde_json::from_str(&device_info)?;
```

File: rust/connlib/clients/android/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/android/src/lib.rs:346]:
```suggestion
    let device_info = serde_json::from_str(&device_info)?;
```

**Review by thomaseizinger — APPROVED** [review]
LGTM

I think we should clean up the error handling but that can happen in a follow-up!

**Comment by jamilbk** [comment]
@conectado Hm I thought we'd cut clients before this got merged. Are the portal changes already deployed to prod? Otherwise we need to revert this, cut clients, and merge this when we can deploy them to prod right?

**Comment by conectado** [comment]
> @conectado Hm I thought we'd cut clients before this got merged. Are the portal changes already deployed to prod? Otherwise we need to revert this, cut clients, and merge this when we can deploy them to prod right?

@jamilbk  We just need to revert the [changes to hashing the device id](https://github.com/firezone/firezone/pull/6857/files#diff-7a253fca7f8a7083ae8f9c7dfba917d6b8d6dad6953f0431b4ae4b078eefc316L48), which we should anyways, that was a left over from the first implementation. Everything else should be harmless as we only add fields not remove it.

**Comment by jamilbk** [comment]
@conectado Got it, thanks for the clarification. If the only breakage here is that it resets verified clients, I'll check prod to see who's using that feature and then just notify them directly.

**Comment by conectado** [comment]
> @conectado Got it, thanks for the clarification. If the only breakage here is that it resets verified clients, I'll check prod to see who's using that feature and then just notify them directly.

@jamilbk yes, that should be it. Though shouldn't we just revert that change? I'm not sure why we hashed the id in the first place but there's no point to making this change anymore

**Comment by thomaseizinger** [comment]
> > @conectado Got it, thanks for the clarification. If the only breakage here is that it resets verified clients, I'll check prod to see who's using that feature and then just notify them directly.
> 
> @jamilbk yes, that should be it. Though shouldn't we just revert that change? I'm not sure why we hashed the id in the first place but there's no point to making this change anymore

I'll send a PR for that so we are unblocked from releasing.

**Comment by conectado** [comment]
> > > @conectado Got it, thanks for the clarification. If the only breakage here is that it resets verified clients, I'll check prod to see who's using that feature and then just notify them directly.
> > 
> > 
> > @jamilbk yes, that should be it. Though shouldn't we just revert that change? I'm not sure why we hashed the id in the first place but there's no point to making this change anymore
> 
> I'll send a PR for that so we are unblocked from releasing.

Thank you @thomaseizinger ! I didn't find the time to sit down and do it myself

**Comment by thomaseizinger** [comment]
PR here: https://github.com/firezone/firezone/pull/6918.

**Comment by jamilbk** [comment]
@conectado Was there a reason why we are not also pulling the SMBIOS UUID for PC systems? That way we have field parity for all desktop systems, similar to what the WARP client pulls.

I left a comment regarding this here: https://github.com/firezone/firezone/issues/6837#issuecomment-2384059326

**Comment by jamilbk** [comment]
Fixed here: https://github.com/firezone/firezone/pull/6921

**Pull request merged**
